yew-functional-macro/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+use proc_macro2::{Span, TokenStream};
+use quote::{format_ident, quote};
+use syn::export::ToTokens;
+use syn::parse::{Parse, ParseStream};
+use syn::{parse_macro_input, Block, FnArg, Ident, Item, ItemFn, Type, Visibility};
+
+struct FunctionalComponent {
+    body: Box<Block>,
+    props_type: Box<Type>,
+    arg: FnArg,
+    vis: Visibility,
+}
+
+impl Parse for FunctionalComponent {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let parsed: Item = input.parse()?;
+
+        match parsed {
+            Item::Fn(func) => {
+                let ItemFn {
+                    attrs: _,
+                    vis,
+                    sig,
+                    block,
+                } = func;
+
+                if !sig.generics.params.is_empty() {
+                    // TODO maybe find a way to handle those
+                    return Err(syn::Error::new_spanned(
+                        sig.generics,
+                        "functional components can't contain generics",
+                    ));
+                }
+
+                if sig.asyncness.is_some() {
+                    return Err(syn::Error::new_spanned(
+                        sig.asyncness,
+                        "functional components can't be async",
+                    ));
+                }
+
+                if sig.constness.is_some() {
+                    return Err(syn::Error::new_spanned(
+                        sig.constness,
+                        "const functions can't be functional components",
+                    ));
+                }
+
+                if sig.abi.is_some() {
+                    return Err(syn::Error::new_spanned(
+                        sig.abi,
+                        "extern functions can't be functional components",
+                    ));
+                }
+
+                let mut inputs = sig.inputs.into_iter();
+                let arg: FnArg = inputs
+                    .next()
+                    .unwrap_or_else(|| syn::parse_quote! { _: &() });
+
+                // Check here so we don't compute anything if params are invalid
+                // `>0` because first one is already consumed.
+                if inputs.len() > 0 {
+                    let params: TokenStream = inputs.map(|it| it.to_token_stream()).collect();
+                    return Err(syn::Error::new_spanned(
+                        params,
+                        "functional components must take only parameter of props",
+                    ));
+                }
+
+                let ty = match arg {
+                    FnArg::Typed(ref arg) => match &*arg.ty {
```

Comment:
You can get rid of the `ref` if you want. I don't really care either way but I've heard many users dislike the `ref` pattern:
```suggestion
                let ty = match &arg {
                    FnArg::Typed(arg) => match &*arg.ty {
```
(Note: you have to modify line `101` to use `&arg` as well)